wgpu-core/src/lib.rs

```diff
@@ -243,6 +243,100 @@ If you are running this program on native and not in a browser and wish to work
 Adapter::downlevel_properties or Device::downlevel_properties to get a listing of the features the current \
 platform supports.";
 
+// #[cfg] attributes in exported macros are interesting!
+//
+// The #[cfg] conditions in a macro's expansion are evaluated using the
+// configuration options (features, target architecture and os, etc.) in force
+// where the macro is *used*, not where it is *defined*. That is, if crate A
+// defines a macro like this:
+//
+//     #[macro_export]
+//     macro_rules! if_bleep {
+//         { } => {
+//             #[cfg(feature = "bleep")]
+//             bleep();
+//         }
+//     }
+//
+// and then crate B uses it like this:
+//
+//     fn f() {
+//         if_bleep! { }
+//     }
+//
+// then it is crate B's `"bleep"` feature, not crate A's, that determines
+// whether the macro expands to a function call or an empty statement. The
+// entire configuration predicate is evaluated in the use's context, not the
+// definition's.
+//
+// Since `wgpu-core` selects back ends using features, we need to make sure the
+// arms of the `gfx_select!` macro are pruned according to `wgpu-core`'s
+// features, not those of whatever crate happens to be using `gfx_select!`. This
+// means we can't use `#[cfg]` attributes in `gfx_select!`s definition itself.
+// Instead, for each backend, `gfx_select!` must use a macro whose definition is
+// selected by `#[cfg]` in `wgpu-core`. The configuration predicate is still
+// evaluated when the macro is used; we've just moved the `#[cfg]` into a macro
+// used by `wgpu-core` itself.
+
+/// Define an exported macro named `$public` that expands to a call if the
+/// configuration predicate `$condition` is true, or to a panic otherwise.
+///
+/// Because of odd technical limitations on exporting macros expanded by other
+/// macros, you must supply a both a public-facing name for the macro and a
+/// private name, which is never used outside this macro. For details:
+/// <https://github.com/rust-lang/rust/pull/52234#issuecomment-976702997>
+macro_rules! define_backend_caller {
+    { $public:ident, $private:ident if $condition:meta } => {
+        #[cfg( $condition )]
+        #[macro_export]
+        macro_rules! $private {
+            ( $backend:literal, $call:expr ) => ( $call )
+        }
+
+        #[cfg(not( $condition ))]
+        #[macro_export]
+        macro_rules! $private {
+            ( $backend:literal, $call:expr ) => (
+                panic!("Unexpected backend {:?}", $backend)
+            )
+        }
+
+        #[doc(hidden)] pub use $private as $public;
```

Comment:
Why do we need the public/private split?

Reply:
That's explained in the doc comment above.